local FlagGrid = require(script.Parent.FlagGrid)
return function()
	local grid: FlagGrid.FlagGrid

	beforeEach(function()
		grid = FlagGrid.new(5, 5, 5)
	end)

	describe("FlagGrid", function()
		it("should give access to its size", function()
			grid = FlagGrid.new(1, 2, 3)
			expect(grid:getXSize()).to.equal(1)
			expect(grid:getYSize()).to.equal(2)
			expect(grid:getZSize()).to.equal(3)
		end)
		it("should set everything to 0 when created", function()
			expect(grid:volumeOr(1, 1, 1, 5, 5, 5)).to.equal(false)
		end)
		it("should set and reset a precise value", function()
			expect(grid:get(2, 2, 2)).to.equal(0)
			grid:set(2, 2, 2)
			expect(grid:get(2, 2, 2)).to.equal(1)
			grid:reset(2, 2, 2)
			expect(grid:get(2, 2, 2)).to.equal(0)
		end)
		it("should set an entire box", function()
			grid:volumeSet(2, 2, 2, 5, 5, 5)
			print(grid)
			expect(grid:volumeAnd(2, 2, 2, 5, 5, 5)).to.equal(true)
		end)
		it("should reset an entire box", function()
			grid:set(3, 3, 3)
			expect(grid:volumeReset(2, 2, 2, 5, 5, 5))
			expect(grid:volumeOr(2, 2, 2, 5, 5, 5)).to.equal(false)
		end)
		it("should perform 'and' upon an entire box", function()
			expect(grid:set(2, 2, 2))
			expect(grid:set(2, 3, 2))
			expect(grid:set(2, 4, 2))
			expect(grid:set(3, 2, 2))
			expect(grid:set(3, 3, 2))
			expect(grid:volumeAnd(2, 2, 2, 3, 4, 2)).to.equal(false)
			expect(grid:set(3, 4, 2))
			expect(grid:volumeAnd(2, 2, 2, 3, 4, 2)).to.equal(true)
		end)
		it("should perform 'or' upon an entire box", function()
			expect(grid:volumeOr(2, 2, 2, 3, 4, 2)).to.equal(false)
			expect(grid:set(3, 4, 2))
			expect(grid:volumeOr(2, 2, 2, 3, 4, 2)).to.equal(true)
			expect(grid:set(3, 3, 2))
			expect(grid:volumeOr(2, 2, 2, 3, 4, 2)).to.equal(true)
		end)
		it("should error if the first position is greater than the second", function()
			expect(function()
				grid:volumeSet(1, 1, 1, 0, 2, 2)
			end).to.throw()
			expect(function()
				grid:volumeReset(1, 1, 1, 0, 2, 2)
			end).to.throw()
			expect(function()
				grid:volumeAnd(1, 1, 1, 0, 2, 2)
			end).to.throw()
			expect(function()
				grid:volumeOr(1, 1, 1, 0, 2, 2)
			end).to.throw()
		end)
		it("should return the first set flag in a line", function()
			grid:set(1, 4, 1)
			expect(grid:getFirstSetInLine(Vector3.new(1, 1, 1), Vector3.new(0, 5, 0))).to.equal(
				Vector3.new(1, 4, 1)
			)
			expect(grid:getFirstSetInLine(Vector3.new(1, 1, 1), Vector3.new(0, 3, 0))).never.to.be.ok()
			expect(grid:getFirstSetInLine(Vector3.new(2, 1, 1), Vector3.new(0, 5, 0))).never.to.be.ok()
		end)
		it("should throw if the length isn't right", function()
			expect(function()
				grid:getFirstSetInLine(Vector3.new(1, 1, 1), Vector3.new(1, 5, 0))
			end).to.throw()
			expect(function()
				grid:getFirstSetInLine(Vector3.new(1, 1, 1), Vector3.new(0, 0, 0))
			end).to.throw()
		end)
	end)
end
